In [1]:
import numpy as np
import pandas as pd

1. Buscar un conjunto de datos a analizar:

Se recomienda que el conjunto de datos a analizar tenga variables numéricas y categóricas. Ya que también se analizarán.

In [2]:
# Buscamos un csv de kaggle
df = pd.read_csv("Datos/cost-of-living.csv")

In [3]:
# Comprobamos las columnas
df.columns

Index(['Unnamed: 0', 'city', 'country', 'x1', 'x2', 'x3', 'x4', 'x5', 'x6',
       'x7', 'x8', 'x9', 'x10', 'x11', 'x12', 'x13', 'x14', 'x15', 'x16',
       'x17', 'x18', 'x19', 'x20', 'x21', 'x22', 'x23', 'x24', 'x25', 'x26',
       'x27', 'x28', 'x29', 'x30', 'x31', 'x32', 'x33', 'x34', 'x35', 'x36',
       'x37', 'x38', 'x39', 'x40', 'x41', 'x42', 'x43', 'x44', 'x45', 'x46',
       'x47', 'x48', 'x49', 'x50', 'x51', 'x52', 'x53', 'x54', 'x55',
       'data_quality'],
      dtype='object')

In [4]:
# Eliminamos la columna Unnamed: 0
df.drop("Unnamed: 0", axis=1, inplace= True)

In [5]:
df.columns

Index(['city', 'country', 'x1', 'x2', 'x3', 'x4', 'x5', 'x6', 'x7', 'x8', 'x9',
       'x10', 'x11', 'x12', 'x13', 'x14', 'x15', 'x16', 'x17', 'x18', 'x19',
       'x20', 'x21', 'x22', 'x23', 'x24', 'x25', 'x26', 'x27', 'x28', 'x29',
       'x30', 'x31', 'x32', 'x33', 'x34', 'x35', 'x36', 'x37', 'x38', 'x39',
       'x40', 'x41', 'x42', 'x43', 'x44', 'x45', 'x46', 'x47', 'x48', 'x49',
       'x50', 'x51', 'x52', 'x53', 'x54', 'x55', 'data_quality'],
      dtype='object')

In [6]:
# Creamos un diccionario para renombrar las columnas a nombres más inteligibles
columnas_nuevas = {"x1":	"Meal", "x2": "Meal for 2 People", "x3":	"McMeal",
"x4":	"Domestic Beer Rest",
"x5":	"Imported Beer Rest",
"x6":	"Cappuccino",
"x7":	"Coke/Pepsi",
"x8":	"Water Rest",
"x9":	"Milk",
"x10": "Bread",
"x11":"Rice",
"x12":"Eggs",
"x13":"Local Cheese",
"x14":	"Chicken Fillets",
"x15":	"Beef Round",
"x16":	"Apples",
"x17":	"Banana",
"x18":	"Oranges",
"x19":	"Tomato",
"x20":	"Potato",
"x21":	"Onion",
"x22":	"Lettuce",
"x23":	"Water",
"x24":	"Bottle of Wine",
"x25":	"Domestic Beer",
"x26":	"Imported Beer",
"x27":	"Cigarettes",
"x28":	"Local Transport",
"x29":	"Monthly Pass",
"x30":	"Taxi Start",
"x31":	"Taxi 1km",
"x32":	"Taxi Waiting",
"x33":	"Gasoline",
"x34":	"Volkswagen Golf",
"x35":	"Toyota Corolla",
"x36":	"Basic" ,
"x37":	"Mobile Tariff",
"x38":	"Internet",
"x39":	"Fitness Club",
"x40":	"Tennis",
"x41":	"Cinema",
"x42":	"Preschool",
"x43":	"Primary School",
"x44":	"Jeans",
"x45":	"Summer Dress",
"x46":	"Running Shoes",
"x47":	"Business Shoes",
"x48":	"1 bedroom Centre",
"x49":	"1 bedroom Out",
"x50":	"3 bedrooms Centre",
"x51":	"3 bedrooms Out",
"x52":	"Price Meter Centre",
"x53":	"Price Meter Out",
"x54":	"Average Monthly",
"x55":	"Mortgage Interest"}

In [7]:
# Modificamos los nombres de las columnas
df.rename(columns= columnas_nuevas, inplace= True)

In [8]:
df.columns

Index(['city', 'country', 'Meal', 'Meal for 2 People', 'McMeal',
       'Domestic Beer Rest', 'Imported Beer Rest', 'Cappuccino', 'Coke/Pepsi',
       'Water Rest', 'Milk', 'Bread', 'Rice', 'Eggs', 'Local Cheese',
       'Chicken Fillets', 'Beef Round', 'Apples', 'Banana', 'Oranges',
       'Tomato', 'Potato', 'Onion', 'Lettuce', 'Water', 'Bottle of Wine',
       'Domestic Beer', 'Imported Beer', 'Cigarettes', 'Local Transport',
       'Monthly Pass', 'Taxi Start', 'Taxi 1km', 'Taxi Waiting', 'Gasoline',
       'Volkswagen Golf', 'Toyota Corolla', 'Basic', 'Mobile Tariff',
       'Internet', 'Fitness Club', 'Tennis', 'Cinema', 'Preschool',
       'Primary School', 'Jeans', 'Summer Dress', 'Running Shoes',
       'Business Shoes', '1 bedroom Centre', '1 bedroom Out',
       '3 bedrooms Centre', '3 bedrooms Out', 'Price Meter Centre',
       'Price Meter Out', 'Average Monthly', 'Mortgage Interest',
       'data_quality'],
      dtype='object')

In [9]:
# Comenzamos a explorar los datos con df.info
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4874 entries, 0 to 4873
Data columns (total 58 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   city                4874 non-null   object 
 1   country             4874 non-null   object 
 2   Meal                4449 non-null   float64
 3   Meal for 2 People   4432 non-null   float64
 4   McMeal              4552 non-null   float64
 5   Domestic Beer Rest  4385 non-null   float64
 6   Imported Beer Rest  4445 non-null   float64
 7   Cappuccino          4540 non-null   float64
 8   Coke/Pepsi          4438 non-null   float64
 9   Water Rest          4571 non-null   float64
 10  Milk                4494 non-null   float64
 11  Bread               4457 non-null   float64
 12  Rice                4538 non-null   float64
 13  Eggs                4369 non-null   float64
 14  Local Cheese        4393 non-null   float64
 15  Chicken Fillets     4318 non-null   float64
 16  Beef R

In [10]:
# Comprobamos el número de nulos por columna
df.isnull().sum()

city                     0
country                  0
Meal                   425
Meal for 2 People      442
McMeal                 322
Domestic Beer Rest     489
Imported Beer Rest     429
Cappuccino             334
Coke/Pepsi             436
Water Rest             303
Milk                   380
Bread                  417
Rice                   336
Eggs                   505
Local Cheese           481
Chicken Fillets        556
Beef Round             562
Apples                 365
Banana                 375
Oranges                435
Tomato                 464
Potato                 430
Onion                  427
Lettuce                495
Water                  384
Bottle of Wine         460
Domestic Beer          725
Imported Beer          651
Cigarettes             485
Local Transport       1473
Monthly Pass          2124
Taxi Start             861
Taxi 1km              1005
Taxi Waiting          1368
Gasoline               578
Volkswagen Golf       1035
Toyota Corolla         970
B

2. Extraer los siguentes estadísiticos del conjunto de datos para varias de las variables disponibles:

Media

Desviación con respecto de la media

Varianza

Desviación típica

Moda

Mediana

Media ponderada

Robustez

Coeficiente de variación

In [11]:
# Media
print("Media del precio del Meal por persona:", df["Meal"].mean())
print("Media del precio del arroz:", df["Rice"].mean())
print("Media del precio del MacMeal:", df["McMeal"].mean())
print("Media del precio de la gasolina:", df["Gasoline"].mean())
print("Media del precio del apartamento:", df["1 bedroom Centre"].mean())
print("Media del precio del apartamento de 3 habitaciones en las afueras:", df["3 bedrooms Out"].mean())

Media del precio del Meal por persona: 10.321643065857495
Media del precio del arroz: 2.182243278977523
Media del precio del MacMeal: 6.8754261862917385
Media del precio de la gasolina: 1.3447672253258844
Media del precio del apartamento: 710.6961562411549
Media del precio del apartamento de 3 habitaciones en las afueras: 973.6038670605612


In [12]:
# Desviación respecto a la media
print("Desviación respecto a la media del precio del Meal por persona:", np.abs(df["Meal"] -df["Meal"].mean()))
print("Desviación respecto a la media del precio del arroz:", np.abs(df["Rice"] -df["Rice"].mean()))
print("Desviación respecto a la media del precio del MacMeal:", np.abs(df["McMeal"] -df["McMeal"].mean()))
print("Desviación respecto a la media del precio de la gasolina:", np.abs(df["Gasoline"] -df["Gasoline"].mean()))
print("Desviación respecto a la media del precio del apartamento:", np.abs(df["1 bedroom Centre"] -df["1 bedroom Centre"].mean()))
print("Desviación respecto a la media del precio del apartamento de 3 habitaciones en las afueras:", np.abs(df["3 bedrooms Out"] -df["3 bedrooms Out"].mean()))

Desviación respecto a la media del precio del Meal por persona: 0        5.421643
1        4.731643
2        7.781643
3        6.781643
4        3.161643
          ...    
4869          NaN
4870          NaN
4871     5.131643
4872    19.458357
4873          NaN
Name: Meal, Length: 4874, dtype: float64
Desviación respecto a la media del precio del arroz: 0       1.302243
1       1.002243
2       1.352243
3       1.292243
4       1.477757
          ...   
4869         NaN
4870         NaN
4871         NaN
4872    4.707757
4873         NaN
Name: Rice, Length: 4874, dtype: float64
Desviación respecto a la media del precio del MacMeal: 0       2.595426
1       1.285426
2       3.375426
3       3.335426
4       0.845426
          ...   
4869         NaN
4870         NaN
4871    6.094574
4872    6.894574
4873    3.674574
Name: McMeal, Length: 4874, dtype: float64
Desviación respecto a la media del precio de la gasolina: 0       0.094767
1       0.174767
2       0.554767
3       0.035233
4    

In [13]:
# Varianza
print("Varianza del precio del Meal por persona:", df["Meal"].var())
print("Varianza del precio del arroz:", df["Rice"].var())
print("Varianza del precio del MacMeal:", df["McMeal"].var())
print("Varianza del precio de la gasolina:", df["Gasoline"].var())
print("Varianza del precio del apartamento:", df["1 bedroom Centre"].var())
print("Varianza del precio del apartamento de 3 habitaciones en las afueras:", df["3 bedrooms Out"].var())

Varianza del precio del Meal por persona: 51.34449556861308
Varianza del precio del arroz: 2.682644029847922
Varianza del precio del MacMeal: 7.123654399736003
Varianza del precio de la gasolina: 0.23576055115839775
Varianza del precio del apartamento: 423025.626860758
Varianza del precio del apartamento de 3 habitaciones en las afueras: 858498.8063168089


In [14]:
# Desviación típica
print("Desviación típica del precio del Meal por persona:", df["Meal"].std())
print("Desviación típica del precio del arroz:", df["Rice"].std())
print("Desviación típica del precio del MacMeal:", df["McMeal"].std())
print("Desviación típica del precio de la gasolina:", df["Gasoline"].std())
print("Desviación típica del precio del apartamento:", df["1 bedroom Centre"].std())
print("Desviación típica del precio del apartamento de 3 habitaciones en las afueras:", df["3 bedrooms Out"].std())

Desviación típica del precio del Meal por persona: 7.165507349002796
Desviación típica del precio del arroz: 1.6378779044385214
Desviación típica del precio del MacMeal: 2.66901749708315
Desviación típica del precio de la gasolina: 0.4855518006952479
Desviación típica del precio del apartamento: 650.4042026776565
Desviación típica del precio del apartamento de 3 habitaciones en las afueras: 926.5521066388058


In [15]:
# Moda
print("Moda del precio del Meal por persona:", df["Meal"].mode()[0])
print("Moda del precio del arroz:", df["Rice"].mode()[0])
print("Moda del precio del MacMeal:", df["McMeal"].mode()[0])
print("Moda del precio de la gasolina:", df["Gasoline"].mode()[0])
print("Moda del precio del apartamento:", df["1 bedroom Centre"].mode()[0])
print("Moda del precio del apartamento de 3 habitaciones en las afueras:", df["3 bedrooms Out"].mode()[0])

Moda del precio del Meal por persona: 15.0
Moda del precio del arroz: 0.93
Moda del precio del MacMeal: 8.0
Moda del precio de la gasolina: 1.23
Moda del precio del apartamento: 623.22
Moda del precio del apartamento de 3 habitaciones en las afueras: 727.09


In [16]:
# Mediana
print("Mediana del precio del Meal por persona:", df["Meal"].median())
print("Mediana del precio del arroz:", df["Rice"].median())
print("Mediana del precio del MacMeal:", df["McMeal"].median())
print("Mediana del precio de la gasolina:", df["Gasoline"].median())
print("Mediana del precio del apartamento:", df["1 bedroom Centre"].median())
print("Mediana del precio del apartamento de 3 habitaciones en las afueras:", df["3 bedrooms Out"].median())

Mediana del precio del Meal por persona: 9.83
Mediana del precio del arroz: 1.68
Mediana del precio del MacMeal: 6.96
Mediana del precio de la gasolina: 1.26
Mediana del precio del apartamento: 519.35
Mediana del precio del apartamento de 3 habitaciones en las afueras: 720.71


In [17]:
# Sacamos w para poder usarlo luego como el peso para la obtención de la media ponderada
w = np.random.randint(0,100, df.shape[0])

In [18]:
w

array([93, 71, 72, ..., 27, 55, 59])

In [19]:
# Media ponderada
print("Media ponderada del precio del Meal por persona:", np.average(df["Meal"].notnull(),weights= w))
print("Media ponderada del precio del arroz:", np.average(df["Rice"].notnull(),weights= w))
print("Media ponderada del precio del MacMeal:", np.average(df["McMeal"].notnull(),weights= w))
print("Media ponderada del precio de la gasolina:", np.average(df["Gasoline"].notnull(),weights= w))
print("Media ponderada del precio del apartamento:", np.average(df["1 bedroom Centre"].notnull(),weights= w))
print("Media ponderada del precio del apartamento de 3 habitaciones en las afueras:", np.average(df["3 bedrooms Out"].notnull(),weights= w))

Media ponderada del precio del Meal por persona: 0.9117422901526352
Media ponderada del precio del arroz: 0.9310205526518543
Media ponderada del precio del MacMeal: 0.9337192595214793
Media ponderada del precio de la gasolina: 0.8806321970722362
Media ponderada del precio del apartamento: 0.7299543968515086
Media ponderada del precio del apartamento de 3 habitaciones en las afueras: 0.6939966266164129


In [20]:
# Coeficiente de variación
print("Coeficiente de variación del precio del Meal por persona:", df["Meal"].std()/df["Meal"].mean())
print("Coeficiente de variación del precio del arroz:", df["Rice"].std()/df["Rice"].mean())
print("Coeficiente de variación del precio del MacMeal:", df["McMeal"].std()/df["Rice"].mean())
print("Coeficiente de variación del precio de la gasolina:", df["Gasoline"].std()/df["Rice"].mean())
print("Coeficiente de variación del precio del apartamento:", df["1 bedroom Centre"].std()/df["Rice"].mean())
print("Coeficiente de variación del precio del apartamento de 3 habitaciones en las afueras:", df["3 bedrooms Out"].std()/df["Rice"].mean())

Coeficiente de variación del precio del Meal por persona: 0.6942215791888077
Coeficiente de variación del precio del arroz: 0.750547805653428
Coeficiente de variación del precio del MacMeal: 1.223061389532015
Coeficiente de variación del precio de la gasolina: 0.2225012240261087
Coeficiente de variación del precio del apartamento: 298.043856495413
Coeficiente de variación del precio del apartamento de 3 habitaciones en las afueras: 424.58699062779846


Robustez

In [21]:
#hacemos una copia para poder trastear con los datos sin alterar el original
copia = df.copy()

La media de la columna "Arroz" es 2.182243278977523

La mediana de la columna "Arroz" es 1.68


In [22]:
#calculamos la media y la mediana de Arroz modificando un dato para ver como de robusto es el dato estadistico
copia["Rice"][0] = copia["Rice"][0] * 1000
copia["Rice"].mean()

/tmp/ipykernel_1824/3232370376.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  copia["Rice"][0] = copia["Rice"][0] * 1000


2.375967386513883

In [23]:
copia["Rice"].median()

1.68

Multiplicamos uno de los valores de la columna Rice por 1000 y comprobamos que la media ha cambiado de 2.1822 a 220555312.89 Por lo tanto deducimos que la media es un valor muy poco robusto ya que al alterar un solo dato ya varia muchisimo el resultado(aun teniendo casi 5000 registros).

Sin embargo la mediana se mantiene igual en los dos casos, 1.68 Por lo que podemos deducir que el valor de la mediana es mucho mas robusto y aunque alteremos un dato no modifica su resultado.

In [24]:
#calculamos punto de ruptura de la media:
med = df["Meal"].mean()
punto_ruptura = 1/med
punto_ruptura

0.09688379976128564

3. Interpretación de los resultados.

Ahora interpreta los resultados obtenidos de los diferentes estadísticos estudiados haciendo hincapié en aspectos interesantes de tus datos y explica lo mejor posible la naturaleza de los mismos.

- **Media**: El valor de la media no nos aporta mucha informacion ya que no es un dato muy robusto, como hemos visto en el ejercicio anterior cualquier dato que esté alejado del resto nos va a alterar mucho el resultado.

- **Mediana**: Como hemos comprobado anteriormente, la mediana sí que es un valor robusto. No varía mucho, a pesar de haber cambiado considerablemente uno de los valores. Lo cual quiere decir que es un valor que nos da información útil.

- Un dato interesante de evaluar es la **desviación típica**. Podemos observar que en artículos como la gasolina, el arroz o el precio de un MacMeal, no hay mucha diferencia de una ciudad a otra. Totalmente opuesto a lo que pasa con la vivienda, en donde existe una gran diferencia de valores. Esto mismo también se puede observar en el **coeficiente de variación**.

In [25]:
#Guardamos el archivo para poder seguir trabajando con el
df_limpio = df.to_csv("Datos/df_limpio.csv")